torchaudio/functional.py

```diff
@@ -1636,3 +1637,80 @@ def detect_pitch_frequency(
     freq = freq.view(shape[:-1] + list(freq.shape[-1:]))
 
     return freq
+
+
+def sliding_window_cmn_internal(
+    waveform: Tensor,
+    cmvn_window: int = 600,
+    min_cmn_window: int = 100,
+    center: bool = False,
+    norm_vars: bool = False,
+) -> Tensor:
+    r"""
+    Apply sliding-window cepstral mean (and optionally variance) normalization per utterance.
+
+    Args:
+        waveform (Tensor): Tensor of audio of dimension (..., freq, time)
+        cmvn_window (int, optional): Window in frames for running average CMN computation (int, default = 600)
+        min_cmn_window (int, optional):  Minimum CMN window used at start of decoding (adds latency only at start). Only applicable if center == false, ignored if center==true (int, default = 100)
+        center (bool, optional): If true, use a window centered on the current frame (to the extent possible, modulo end effects). If false, window is to the left. (bool, default = false)
+        norm_vars (bool, optional): If true, normalize variance to one. (bool, default = false)
+
+    Returns:
+        Tensor: Tensor of freq of dimension (..., frame)
+    """
+    last_window_start = last_window_end = -1
+    num_frames, num_feats = waveform.shape
+    cur_sum = torch.zeros(num_feats)
```

Comment:
FYI: To make it device (CUDA) compatible, can you give `dtype` and `device` to all the temporary Tensors created in the function?

e.g. 
```python
dtype = waveform.dtype
device = waveform.device

cur_sum = torch.zeros(num_feats, dtype=dtype, device=device)

...

cmvn_waveform = torch.zeros(num_frames, num_feats, dtype=dtype, device=device)

... 
etc
```

Reply:
ths